### PR TITLE
feat: expose IResizeLayer in dynamo.conversion.impl

### PR DESCRIPTION
**Pull request by bowang007**
# Description

#2219 

## Type of change
Expose IResizeLayer in dynamo.conversion.impl

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/impl/interpolate.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    align_corners: Optional[bool],
+) -> TRTTensor:
+    resize_layer = ctx.net.add_resize(input)
+    # output size calculation
+    # Pytorch assures that only one of out_shape/scale_factor is not None
+    # Pytorch assures that dimensions match for out_shape/scale factor
+    if out_shape is not None:
+        resize_layer.shape = list(input.shape)[:2] + list(out_shape)
+    else:
+        resize_layer.scales = [1, 1] + list(scale_factor)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/interpolate.py:29]: Should both branches of the conditional ensure that both `shape` and `scales` are set, or can TRT handle cases where only one is set?

File: py/torch_tensorrt/dynamo/conversion/impl/interpolate.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    # interpolate mode
+    if interpolate_mode == "nearest" or None:
+        resize_layer.resize_mode = trt.ResizeMode.NEAREST
+    elif interpolate_mode == "bilinear" or "linear":
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/interpolate.py:34]: Rewrite as:
```python
elif interpolate_mode in ("bilinear", "linear")
```

File: py/torch_tensorrt/dynamo/conversion/impl/interpolate.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    # output size calculation
+    # Pytorch assures that only one of out_shape/scale_factor is not None
+    # Pytorch assures that dimensions match for out_shape/scale factor
+    if out_shape is not None:
+        resize_layer.shape = list(input.shape)[:2] + list(out_shape)
+    else:
+        resize_layer.scales = [1, 1] + list(scale_factor)
+
+    # interpolate mode
+    if interpolate_mode == "nearest" or None:
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/interpolate.py:32]: Rewrite as:
```python
if interpolate_mode == "nearest" or interpolate_mode is None
```

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 4]
```diff
@@ -2026,3 +2026,24 @@ def aten_ops_addmm(
...
+        ctx,
+        target,
+        SourceIR.ATEN,
+        name,
+        input=args[0],
+        out_shape=kwargs.get("size", 1),
+        scale_factor=kwargs.get("scale_factor", 1),
+        interpolate_mode=kwargs.get("mode", 1),
+        align_corners=kwargs.get("align_corners", 1),
+    )
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:2049]: The decorator's `key` shoud exactly match what the schema has, which typically is something like `torch.ops.aten.xxxxx`. And the type of arguments should also exactly match with the schema. After the corresponding changes, consider using `args_bounds_check(args, ...)` that takes arguments out of `args` instead of `kwargs`.

File: tests/py/dynamo/conversion/test_interpolate.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
+import torch
+from parameterized import parameterized
+from torch.testing._internal.common_utils import run_tests
+
+from .harness import DispatchTestCase
+
+
+class TestNearestInterpolateConverter(DispatchTestCase):
```
Comment (zewenli98) [inline tests/py/dynamo/conversion/test_interpolate.py:8]: Consider adding more diverse test cases, like data with different dimensions.

**Review by zewenli98 — DISMISSED** [review]
Our converters come from the [schema](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml). The decorator's `key` should be one of the scheme's functions.

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 4 above)
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: from my understanding, when we have some function schema as: 
`
call_function[target=torch.ops.aten.upsample_bilinear2d.vec](args = (%input_1,), kwargs = {output_size: (4, 4), align_corners: True, scale_factors: None})
`
we should use `kwargs` since arguments like `output_size` are in kwargs instead, right?

Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: The reason why `output_size` and so on appeared in `kwargs` is that you passed in arguments in tests by: 
```
return torch.ops.aten.upsample_nearest2d.vec(
    input, output_size=output_shape, scale_factors=None
)
```
If you remove `=`, as you can see we don't use `=` in other tests, they should be in `args`. 

For the underlying reason, let's go back to the schema:
```
func: upsample_bilinear2d.vec(Tensor input, SymInt[]? output_size, bool align_corners, float[]? scale_factors) -> Tensor
```
They should belong to `args`, instead of `kwargs`. 
Only those arguments following `*` are considered as `kwargs`. For example,
```
replication_pad3d.out(Tensor self, SymInt[6] padding, *, Tensor(a!) out) -> Tensor(a!)
```
Based on the schema above, `self` and `padding` should be considered as `args` while `out` should be `kwargs`. The use of `*` in the function signature enforces this syntax.

I encountered this problem before as well, @gs-olive told me **"The current tracer being used by default in testing is a very barebones tracer, so it knows nothing about `aten` ops specifically. It seems that if an argument is specified with `=`, it will consider that a `kwarg`"**

Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py]: cool, thanks!

File: py/torch_tensorrt/dynamo/conversion/impl/upsample.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        raise RuntimeError(
+            f"At least one of out_shape and scale_factors should be specified."
+        )
+
+    # interpolate mode
+    if resize_mode == "nearest" or None:
+        resize_layer.resize_mode = trt.ResizeMode.NEAREST
+    elif resize_mode == "bilinear":
+        resize_layer.resize_mode = trt.ResizeMode.LINEAR
+        if align_corners is None or align_corners is False:
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/upsample.py:40]: Switch `align_corners is False` to `not align_corners`

File: py/torch_tensorrt/dynamo/conversion/impl/upsample.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+def upsample(
+    ctx: ConversionContext,
+    target: Target,
+    source_ir: Optional[SourceIR],
+    name: str,
+    input: TRTTensor,
+    out_shape: Optional[Sequence[int]],
+    scale_factors: Optional[Sequence[float]],
+    resize_mode: str,
+    align_corners: Optional[bool],
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/upsample.py:20]: This argument could be switched to just `bool`, since `None` and `False` seem to have the same meaning in the function

File: py/torch_tensorrt/dynamo/conversion/impl/upsample.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+) -> TRTTensor:
+    resize_layer = ctx.net.add_resize(input)
+    # output size calculation
+    # Pytorch assumes that one of out_shape/scale_factor is None
+    # Pytorch assumes that dimensions match for out_shape/scale factor
+    if out_shape is not None:
+        resize_layer.shape = list(input.shape)[:2] + list(out_shape)
+    elif scale_factors is not None:
+        resize_layer.scales = [1.0, 1.0] + list(scale_factors)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/upsample.py:29]: Are these specific to the 2D case? If so, could they be generalized for 1D or 3D as well?

**Review by gs-olive — COMMENTED** [review]
Added a few suggestions, but overall looks good! Could you also add support for the non-`vec` variants, such as [`upsample_bilinear2d`](https://github.com/pytorch/pytorch/blob/4c5fe66880031f9acee4622e1d085f5ca4a470f6/aten/src/ATen/native/native_functions.yaml#L12368).

File: py/torch_tensorrt/dynamo/conversion/impl/upsample.py (hunk 8 above)
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/impl/upsample.py:29]: `upsample` or `interpolate` functions in pytorch assumes that the input is an image with (n, c, h, w ...), in other words, the input of these functions should at least be `dim == 3` since the first 2 dimensions are `batchsize` and `channels`.

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/upsample.py:29]: Understood - thanks for the clarification, that makes sense

File: py/torch_tensorrt/dynamo/conversion/impl/interpolate.py (hunk 1 above)
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/impl/interpolate.py]: @gs-olive 
Looks like there are some internal assertions in Pytorch that checks the arguments for `shape` and `scales`. 
So pytorch will throws out an error if none of them is set or both of them are set. 
Since this is a converter so we assumes that the arguments are valid or pytorch will throw out error right? 
Please correct me if I'm wrong, thanks!

File: py/torch_tensorrt/dynamo/conversion/impl/upsample.py (hunk 7 above)
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/impl/upsample.py:20]: Thanks this makes sense!

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 9]
```diff
@@ -2026,3 +2026,45 @@ def aten_ops_addmm(
...
+        name,
+        input=args[0],
+        out_shape=args_bounds_check(args, 1),
+        scale_factors=args_bounds_check(args, 2),
+        resize_mode="nearest",
+        align_corners=None,
+    )
+
+
+@dynamo_tensorrt_converter(torch.ops.aten.upsample_bilinear2d.vec)  # type:ignore[misc]
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:2052]: The `# type:ignore[misc]` can be removed for these decorators. If linting shows issues with this, adding `--no-verify` to `git commit` will allow you to bypass the issues.

**Review by gs-olive — COMMENTED** [review]
Looks good overall - added minor comment about `mypy` linting. Also needs a rebase onto the latest `main` to resolve merge conflicts

**Comment by bowang007** [comment]
> @bowang007 - actually, I realized the [`upsample_bilinear2d`](https://github.com/pytorch/pytorch/blob/4c5fe66880031f9acee4622e1d085f5ca4a470f6/aten/src/ATen/native/native_functions.yaml#L12368) op is not in the Core ATen IR set, so it is okay to omit it. That error arises when the decorator specified for the converter is something like `aten.upsample_bilinear2d` instead of `aten.upsample_bilinear2d.default`

In that case, how can we add converter for `aten.upsample_bilinear2d`?

**Comment by gs-olive** [comment]
We can add this converter using the decorator `@torch.ops.aten.upsample_bilinear2d.default` as opposed to `@torch.ops.aten.upsample_bilinear2d`.

**Comment by bowang007** [comment]
> Our converters come from the [schema](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml). The decorator's `key` should be one of the scheme's functions.

yeah that makes sense! Thanks!

**Review by gs-olive — APPROVED** [review]
Just needs a rebase, then looks good to me
